pkg/capi/aws.go

```diff
@@ -341,3 +338,84 @@ func newAWSMachineTemplate(name string, mapiProviderSpec *mapiv1.AWSMachineProvi
 
 	return awsMachineTemplate
 }
+
+// createCAPIMachineSetWithRetry creates a CAPI MachineSet with retry logic for capacity constraints.
+// It tries different instance types when encountering insufficient capacity errors.
+func createCAPIMachineSetWithRetry(ctx context.Context, cl client.Client, machineSetName string, clusterName string, platform configv1.PlatformType, mapiDefaultProviderSpec *mapiv1.AWSMachineProviderConfig, templateConfigurator func(*awsv1.AWSMachineTemplate, string)) (*clusterv1.MachineSet, *awsv1.AWSMachineTemplate) {
```

Comment:
Have you somehow been able to test this works as intended in case the instance is not available?

Reply:
I haven't managed to test that, because I run many times on my local but didn't meet InsufficientInstanceCapacity.
I was hoping to meet InsufficientInstanceCapacit error in CI, but seems [it](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-api-actuator-pkg/429/pull-ci-openshift-cluster-api-actuator-pkg-master-e2e-aws-capi-techpreview/1963866020871409664) haven't. I will try in my local again.

Reply:
OK SGTM